services/route_manager/src/main/java/com/futurewei/alcor/route/controller/NeutronRouterController.java

```diff
@@ -0,0 +1,363 @@
+                    BeanUtils.copyProperties(routerExtraAttribute, neutronRouterWebRequestObject);
+                }
+
+                neutronRouters.add(neutronRouterWebRequestObject);
+            }
+
+        } catch (ParameterNullOrEmptyException e) {
+            throw new Exception(e);
+        } catch (ResourceNotFoundException e) {
+            throw new Exception(e);
+        }
+
+        return new NeutronRoutersWebJson(neutronRouters);
+    }
+
+    /**
+     * Create a Neutron router
+     * @param projectid
+     * @param resource
+     * @return
+     * @throws Exception
+     */
+    @RequestMapping(
+            method = POST,
+            value = {"/project/{projectid}/routers"})
+    @DurationStatistics
+    public NeutronRouterWebJson createNeutronRouters(@PathVariable String projectid, @RequestBody NeutronRouterWebJson resource) throws Exception {
+
+        NeutronRouterWebRequestObject neutronRouterWebRequestObject = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+
+            // check resource
+            if (!RouteManagerUtil.checkNeutronRouterWebResourceIsValid(resource)) {
+                throw new ResourceNotValidException("request resource is invalid");
+            }
+
+            neutronRouterWebRequestObject = resource.getRouter();
+            String id = neutronRouterWebRequestObject.getId();
+
+            if (id == null || StringUtils.isEmpty(id)) {
+                UUID vpcId = UUID.randomUUID();
+                neutronRouterWebRequestObject.setId(vpcId.toString());
+            }
+            RestPreconditionsUtil.verifyResourceNotNull(neutronRouterWebRequestObject);
+
+            // configure default value
+            neutronRouterWebRequestObject = RouteManagerUtil.configureNeutronRouterParameters(neutronRouterWebRequestObject);
+
+            // save router and router_extra_attribute
+            this.neutronRouterService.saveRouterAndRouterExtraAttribute(neutronRouterWebRequestObject);
+
+        } catch (Exception e) {
+            throw e;
+        }
+
+        return new NeutronRouterWebJson(neutronRouterWebRequestObject);
+    }
+
+    /**
+     * Update a Neutron router
+     * @param projectid
+     * @param routerid
+     * @param resource
+     * @return
+     * @throws Exception
+     */
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/routers/{routerid}"})
+    @DurationStatistics
+    public NeutronRouterWebJson updateNeutronRouterByRouterId(@PathVariable String projectid,@PathVariable String routerid, @RequestBody NeutronRouterWebJson resource) throws Exception {
+        NeutronRouterWebRequestObject neutronRouterWebRequestObject = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+
+            // check resource
+            if (!RouteManagerUtil.checkNeutronRouterWebResourceIsValid(resource)) {
```

Comment:
what if the resource.getid() != routerid?

Reply:
User are not allowed to change the router_id